### ISSUE TITLE
FYI: Emscripten changes `long double` alignment in 2.0.26

### ISSUE THREAD

**Issue by kleisauke** [hoodmane/libffi-emscripten#5]
Quoting from `ChangeLog.md`:
> The alignment of `long double`, which is a 128-bit floating-point value implemented in software, is reduced from 16 to 8. The lower alignment allows `max_align_t` to properly match the alignment we use for malloc, which is 8 (raising malloc's alignment to achieve correctness the other way would come with a performance regression). (#10072)

I think this patch can be applied once Emscripten 2.0.26 is released:
```diff
--- a/src/wasm32/ffi.c
+++ b/src/wasm32/ffi.c
@@ -276,7 +276,7 @@ ffi_call, (ffi_cif * cif, ffi_fp fn, void *rvalue, void **avalue),
         DEREF_U32(varargs_addr, 1) = DEREF_U32(arg_ptr, 1);
         break;
       case FFI_TYPE_LONGDOUBLE:
-        STACK_ALLOC(varargs_addr, 16, 16);
+        STACK_ALLOC(varargs_addr, 16, 8);
         DEREF_U32(varargs_addr, 0) = DEREF_U32(arg_ptr, 0);
         DEREF_U32(varargs_addr, 1) = DEREF_U32(arg_ptr, 1);
         DEREF_U32(varargs_addr, 2) = DEREF_U32(arg_ptr, 1);
@@ -540,7 +540,7 @@ ffi_prep_closure_loc_helper,
         STORE_U64(cur_ptr, 0, cur_arg);
         break;
       case FFI_TYPE_LONGDOUBLE:
-        STACK_ALLOC(cur_ptr, 16, 16);
+        STACK_ALLOC(cur_ptr, 16, 8);
         DEREF_U32(args_ptr, carg_idx) = cur_ptr;
         STORE_U64(cur_ptr, 0, cur_arg);
         cur_arg = args[jsarg_idx++];
```

A quick check with the "tip-of-tree" build of emsdk looks like the libffi test suite is unaffected (i.e. the same number of expected passes and unexpected failures), both with and without this patch, so perhaps it doesn't matter that much.

**Comment by hoodmane** [comment]
Well Pyodide is currently still on Emscripten 2.0.16 (CI is failing on update to v2.0.24). I'd want to wait until Pyodide upgrades.

**Comment by hoodmane** [comment]
> A quick check with the "tip-of-tree" build of emsdk looks like the libffi test suite is unaffected (i.e. the same number of expected passes and unexpected failures), both with and without this patch, so perhaps it doesn't matter that much.

I'm surprised by that, could every long double in the test suite be 16 byte aligned by bad luck?

**Comment by kleisauke** [comment]
> I'm surprised by that, could every long double in the test suite be 16 byte aligned by bad luck?

I'm not sure, I haven't investigated it further yet. FWIW, the Node CircleCI builds at commit https://github.com/hoodmane/libffi-emscripten/commit/02d20c875dd8edf4033ed0578c5865e2e074a612 uses Emscripten 2.0.16.

**Comment by hoodmane** [comment]
Closed by 2f79f71495792a8b0fa6aa22ef09ddb22f1fce2a
